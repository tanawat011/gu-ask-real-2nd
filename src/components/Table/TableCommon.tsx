import tw, { styled } from 'twin.macro'

import { DARK_THEME, LIGHT_THEME } from 'constants/twTheme'

type Columns<T> = {
  key: keyof T
  label: string
}

type TwTableCommonProps<T> = {
  columns: Columns<T>[]
  data: T[]
}

const TwTable = tw.table`w-full`

const TwThead = styled.thead(() => [
  [DARK_THEME.TABLE.HEADER.BG_COLOR],
  [LIGHT_THEME.TABLE.HEADER.BG_COLOR],
])
const TwTheadTh = styled.th(() => [
  [DARK_THEME.TABLE.HEADER.HEAD.TEXT_COLOR],
  [LIGHT_THEME.TABLE.HEADER.HEAD.TEXT_COLOR],
  tw`px-6 py-3 text-left font-semibold text-xs`,
])

const TwTbody = tw.tbody``
const TwTr = styled.tr(() => [
  [
    DARK_THEME.TABLE.BODY.ROW.BG_COLOR,
    DARK_THEME.TABLE.BODY.ROW.BORDER_COLOR,
    DARK_THEME.TABLE.BODY.ROW.HOVER.BG_COLOR,
  ],
  [
    LIGHT_THEME.TABLE.BODY.ROW.BG_COLOR,
    LIGHT_THEME.TABLE.BODY.ROW.BORDER_COLOR,
    LIGHT_THEME.TABLE.BODY.ROW.HOVER.BG_COLOR,
  ],
  tw`border-t`,
])
const TwTd = styled.td(() => [
  [DARK_THEME.TABLE.BODY.DATA.TEXT_COLOR],
  [LIGHT_THEME.TABLE.BODY.DATA.TEXT_COLOR],
  tw`px-6 py-4 font-semibold text-sm leading-6`,
])

export const TableCommon = <T,>({ columns, data }: TwTableCommonProps<T>) => {
  return (
    <TwTable>
      <TwThead>
        <tr>
          {columns.map((column, i) => (
            <TwTheadTh key={i}>{column.label}</TwTheadTh>
          ))}
        </tr>
      </TwThead>

      <TwTbody>
        {data.map((row, i) => (
          <TwTr key={i}>
            {columns.map((column, j) => (
              <TwTd key={j}>{row[column.key as never]}</TwTd>
            ))}
          </TwTr>
        ))}
      </TwTbody>
    </TwTable>
  )
}
