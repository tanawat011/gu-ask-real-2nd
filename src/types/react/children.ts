export type Children = { children: JSX.Element }
