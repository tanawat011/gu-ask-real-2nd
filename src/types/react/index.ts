export * from './children'
export * from './reactRouterDom'
