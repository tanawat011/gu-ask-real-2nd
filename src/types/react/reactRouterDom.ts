import type { Scroll } from 'types/common'

export type OutletContextProps = {
  scroll: Scroll
}
