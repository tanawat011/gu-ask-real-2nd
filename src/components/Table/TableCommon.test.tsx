import { render } from '@testing-library/react'

import { TableCommon } from './TableCommon'

jest.mock('./TableCommon', () => jest.requireActual('./TableCommon'))

describe('<TableCommon />', () => {
  test('renders correctly', () => {
    expect.assertions(1)

    const { asFragment } = render(
      <TableCommon
        columns={[
          { key: 'name', label: 'Name' },
          { key: 'age', label: 'Age' },
        ]}
        data={[
          { name: 'John', age: 20 },
          { name: 'Jane', age: 21 },
        ]}
      />,
    )

    expect(asFragment()).toMatchSnapshot()
  })
})
