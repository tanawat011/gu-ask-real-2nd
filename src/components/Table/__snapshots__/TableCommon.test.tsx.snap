// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableCommon /> renders correctly 1`] = `
<DocumentFragment>
  .c0 {
  width: 100%;
}

.c2 {
  --tw-bg-opacity: 1;
  background-color: rgb(249 250 251 / var(--tw-bg-opacity));
}

.dark .c1 {
  --tw-bg-opacity: 1;
  background-color: rgb(55 65 81 / var(--tw-bg-opacity));
}

.c4 {
  --tw-text-opacity: 1;
  color: rgb(107 114 128 / var(--tw-text-opacity));
  padding-left: 1.5rem;
  padding-right: 1.5rem;
  padding-top: 0.75rem;
  padding-bottom: 0.75rem;
  text-align: left;
  font-size: 0.75rem;
  line-height: 1rem;
  font-weight: 600;
}

.dark .c3 {
  --tw-text-opacity: 1;
  color: rgb(255 255 255 / var(--tw-text-opacity));
}

.c6 {
  --tw-bg-opacity: 1;
  background-color: rgb(255 255 255 / var(--tw-bg-opacity));
  --tw-border-opacity: 1;
  border-color: rgb(229 231 235 / var(--tw-border-opacity));
  border-top-width: 1px;
}

.dark .c5 {
  --tw-bg-opacity: 1;
  background-color: rgb(31 41 55 / var(--tw-bg-opacity));
}

.dark .c5 {
  --tw-border-opacity: 1;
  border-color: rgb(55 65 81 / var(--tw-border-opacity));
}

.dark .c5:hover {
  --tw-bg-opacity: 0.5;
  background-color: rgb(55 65 81 / var(--tw-bg-opacity));
}

.c6:hover {
  --tw-bg-opacity: 0.5;
  background-color: rgb(243 244 246 / var(--tw-bg-opacity));
}

.c8 {
  --tw-text-opacity: 1;
  color: rgb(156 163 175 / var(--tw-text-opacity));
  padding-left: 1.5rem;
  padding-right: 1.5rem;
  padding-top: 1rem;
  padding-bottom: 1rem;
  font-size: 0.875rem;
  line-height: 1.5rem;
  font-weight: 600;
}

.dark .c7 {
  --tw-text-opacity: 1;
  color: rgb(156 163 175 / var(--tw-text-opacity));
}

<table
    class="c0"
  >
    <thead
      class="c1 c2"
    >
      <tr>
        <th
          class="c3 c4"
        >
          Name
        </th>
        <th
          class="c3 c4"
        >
          Age
        </th>
      </tr>
    </thead>
    <tbody
      class=""
    >
      <tr
        class="c5 c6"
      >
        <td
          class="c7 c8"
        >
          John
        </td>
        <td
          class="c7 c8"
        >
          20
        </td>
      </tr>
      <tr
        class="c5 c6"
      >
        <td
          class="c7 c8"
        >
          Jane
        </td>
        <td
          class="c7 c8"
        >
          21
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
